cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
   friend array<Y> const_array_cast(const array<U>&);
 
   public:
-    explicit array(std::int64_t element_count)
-        : data_(new T[element_count]), // TODO: Use custom oneDAL allocator
-          size_(element_count) {}
-
-    explicit array(const detail::shared<T>& data, std::int64_t size)
-        : data_(data),
-          size_(size) {}
+    array()
+        : data_owned_ptr_(nullptr),
+          size_(0),
+          capacity_(0) {}
+
+    explicit array(std::int64_t element_count, T element = {})
+        : data_owned_ptr_(new T[element_count], // TODO: Use custom oneDAL allocator
+                [](T* obj){ delete[] obj; }),
```

Comment:
I know, this is prototype, for production it makes sense to replace those lambdas with functors like `std::default_delete` or `empty_deleter`.